File: python/lib/orchestrator.py
#!/usr/bin/env python3
"""
Main orchestrator for the GitHub Issue Orchestrator.

This module coordinates all the components to process GitHub issues
with Claude Code in a concurrent, stateful manner.
"""

import asyncio
import logging
from datetime import datetime
from pathlib import Path

import aiohttp

from .config import Config
from .github_client import GitHubClient
from .models import IssueState, ProcessStatus
from .repo_manager import RepoManager
from .state_manager import StateManager
from .telegram_notifier import TelegramNotifier
from .slack_notifier import SlackNotifier
from .claude_processor import ClaudeProcessor


class IssueOrchestrator:
    """Main orchestrator for processing GitHub issues"""
    
    def __init__(self):
        self.config = Config()
        self.state_manager = StateManager()
        # Create a generic GitHub client without a specific repo
        self.github_client = GitHubClient(
            self.config.github_token
        )
        # Initialize notifiers list
        self.notifiers = []
        
        # Add Telegram notifier if configured
        if self.config.telegram_bot_token and self.config.telegram_chat_id:
            self.notifiers.append(TelegramNotifier(
                self.config.telegram_bot_token,
                self.config.telegram_chat_id
            ))
        
        # Add Slack notifier if configured
        if self.config.slack_bot_token and self.config.slack_channel_id:
            self.notifiers.append(SlackNotifier(
                self.config.slack_bot_token,
                self.config.slack_channel_id
            ))
        
        # For backward compatibility, keep single notifier reference
        self.notifier = self.notifiers[0] if self.notifiers else None
        self.repo_manager = RepoManager(self.config)
        self.processor = ClaudeProcessor(self.config, self.notifiers, self.repo_manager)
        
        # Setup logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(message)s',
            handlers=[
                logging.FileHandler('orchestrator.log'),
                logging.StreamHandler()
            ]
        )
    
    async def run(self):
        """Main orchestration loop"""
        logging.info("Starting Issue Orchestrator")
        
        # Validate base repo paths exist for all repos
        for repo_config in self.config.github_repos:
            base_path = Path(repo_config["base_repo_path"]).expanduser().resolve()
            if not base_path.exists():
                logging.info(f"Creating base repo path: {base_path}")
                base_path.mkdir(parents=True, exist_ok=True)
            
            if not base_path.is_dir():
                logging.error(f"Base repo path is not a directory: {base_path}")
                raise NotADirectoryError(f"Base repo path is not a directory: {base_path}")
            
            logging.info(f"Using base repo path for {repo_config['name']}: {base_path}")
        
        async with aiohttp.ClientSession() as session:
            try:
                # Get ready issues from all configured repos
                all_issues = []
                for repo_config in self.config.github_repos:
                    repo_name = repo_config["name"]
                    try:
                        repo_issues = await self.github_client.get_ready_issues(session, repo_name)
                        logging.info(f"Found {len(repo_issues)} ready issues in {repo_name}")
                        all_issues.extend(repo_issues)
                    except Exception as e:
                        logging.error(f"Failed to get issues from {repo_name}: {e}")
                        continue
                
                issues = all_issues
                logging.info(f"Found {len(issues)} total ready issues across all repos")
                
                # Filter out already processing issues
                active_issues = self.state_manager.get_active_issues()
                new_issues = [
                    issue for issue in issues 
                    if issue["number"] not in active_issues
                ]
                
                # Process issues up to max concurrent limit
                current_processing = len(active_issues)
                available_slots = self.config.max_concurrent - current_processing
                
                if available_slots > 0 and new_issues:
                    issues_to_process = new_issues[:available_slots]
                    
                    # Create tasks for concurrent processing
                    tasks = []
                    for issue in issues_to_process:
                        # Get an available agent index
                        agent_index = self.state_manager.get_available_agent_index(self.config.max_concurrent)
                        if agent_index is None:
                            logging.warning(f"No available agent slots for issue #{issue['number']}. Skipping.")
                            continue
                        
                        # Reserve the agent index immediately by creating a placeholder state
                        # This prevents race conditions where multiple issues get the same agent
                        placeholder_state = IssueState(
                            issue_number=issue["number"],
                            status=ProcessStatus.RUNNING.value,
                            session_id=None,
                            branch=f"issue-{issue['number']}",
                            start_time=datetime.now().isoformat(),
                            agent_index=agent_index,
                            repo_name=issue.get('repo_name')  # Store repo name
                        )
                        self.state_manager.set_state(issue["number"], placeholder_state)
                        
                        task = asyncio.create_task(
                            self._process_issue(session, issue, agent_index)
                        )
                        tasks.append(task)
                    
                    # Wait for all tasks to complete
                    await asyncio.gather(*tasks, return_exceptions=True)
                
                # Save state
                await self.state_manager.save_states()
                
            except Exception as e:
                logging.error(f"Orchestrator error: {e}")
                raise
    
    async def _process_issue(self, session: aiohttp.ClientSession, issue: dict, agent_index: int):
        """Process a single issue"""
        issue_number = issue["number"]
        issue_title = issue["title"]
        repo_name = issue.get('repo_name')
        
        try:
            logging.info(f"Starting to process issue #{issue_number} from {repo_name} with agent {agent_index}")
            
            # Update labels: remove 'ready-for-claude', add 'claude-working'
            await self.github_client.update_issue_labels(
                session,
                issue_number,
                add_labels=["claude-working"],
                remove_labels=["ready-for-claude"],
                repo=repo_name
            )
            
            # Update the placeholder state with additional information
            # The state was already created in the main loop to reserve the agent index
            repo_path = self.config.get_repo_path(agent_index, repo_name)
            
            state = self.state_manager.get_state(issue_number)
            if not state:
                # Fallback: create state if it doesn't exist (shouldn't happen normally)
                state = IssueState(
                    issue_number=issue_number,
                    status=ProcessStatus.RUNNING.value,
                    session_id=None,
                    branch=f"issue-{issue_number}",
                    start_time=datetime.now().isoformat(),
                    agent_index=agent_index,
                    repo_name=repo_name
                )
                self.state_manager.set_state(issue_number, state)
            
            await self.state_manager.save_states()
            
            # Send start notification to all notifiers
            for notifier in self.notifiers:
                try:
                    if isinstance(notifier, SlackNotifier):
                        await notifier.notify_start(issue_number, issue_title, repo_name)
                    else:
                        await notifier.notify_start(issue_number, issue_title)
                except Exception as e:
                    logging.error(f"Failed to send start notification: {e}")
            
            # Process with Claude
            result, session_id = await self.processor.process_issue(issue_number, agent_index, self.state_manager, repo_name)
            
            # Update state based on result (session_id already updated in read_output)
            state = self.state_manager.get_state(issue_number)  # Get latest state with session_id
            if state:
                state.status = result.value
                state.end_time = datetime.now().isoformat()
                
                # Save updated state
                self.state_manager.set_state(issue_number, state)
                await self.state_manager.save_states()
            
            if result == ProcessStatus.COMPLETED:
                # Calculate duration
                start = datetime.fromisoformat(state.start_time)
                end = datetime.fromisoformat(state.end_time)
                duration = str(end - start).split('.')[0]
                
                # Send completion notification to all notifiers
                for notifier in self.notifiers:
                    try:
                        if isinstance(notifier, SlackNotifier):
                            await notifier.notify_complete(issue_number, duration, repo_name)
                        else:
                            await notifier.notify_complete(issue_number, duration)
                    except Exception as e:
                        logging.error(f"Failed to send completion notification: {e}")
                
                # Update labels: remove 'claude-working', add 'claude-completed'
                await self.github_client.update_issue_labels(
                    session,
                    issue_number,
                    add_labels=["claude-completed"],
                    remove_labels=["claude-working"],
                    repo=repo_name
                )
                
                # Remove from state manager
                self.state_manager.remove_state(issue_number)
                
            elif result == ProcessStatus.NEEDS_INPUT:
                # Keep in state, notify user
                for notifier in self.notifiers:
                    try:
                        if isinstance(notifier, SlackNotifier):
                            await notifier.notify_needs_input(
                                issue_number,
                                state.last_output or "No output available",
                                repo_name
                            )
                        else:
                            await notifier.notify_needs_input(
                                issue_number,
                                state.last_output or "No output available"
                            )
                    except Exception as e:
                        logging.error(f"Failed to send needs input notification: {e}")
                
            elif result == ProcessStatus.FAILED:
                # Update labels: remove 'claude-working', add 'claude-failed'
                await self.github_client.update_issue_labels(
                    session,
                    issue_number,
                    add_labels=["claude-failed"],
                    remove_labels=["claude-working"],
                    repo=repo_name
                )
                
                # Remove from state manager
                self.state_manager.remove_state(issue_number)
            
            await self.state_manager.save_states()
            
        except Exception as e:
            logging.error(f"Error processing issue #{issue_number} with agent {agent_index}: {e}")
            for notifier in self.notifiers:
                try:
                    if isinstance(notifier, SlackNotifier):
                        await notifier.notify_error(issue_number, str(e), None, repo_name)
                    else:
                        await notifier.notify_error(issue_number, str(e))
                except Exception as notify_error:
                    logging.error(f"Failed to send error notification: {notify_error}")
            
            # Try to update labels to indicate failure
            try:
                await self.github_client.update_issue_labels(
                    session,
                    issue_number,
                    add_labels=["claude-failed"],
                    remove_labels=["claude-working", "ready-for-claude"],
                    repo=repo_name
                )
            except:
                pass
            
            # Remove from state manager
            self.state_manager.remove_state(issue_number)
            await self.state_manager.save_states()


async def main():
    """Main entry point"""
    orchestrator = IssueOrchestrator()
    await orchestrator.run()


if __name__ == "__main__":
    asyncio.run(main())

File: src/lib/orchestrator.ts
import { mkdir } from "fs/promises";
import { resolve } from "path";
import { Config, ProcessorName, SUPPORTED_PROCESSORS } from "./config";
import { buildStartupBanner } from "./startupBanner";
import { GitHubClient, GitHubIssue } from "./githubClient";
import { IssueState, ProcessStatus } from "./models";
import { RepoManager } from "./repoManager";
import { StateManager } from "./stateManager";
import { ClaudeProcessor } from "./processors/claude";
import { CodexProcessor } from "./processors/codex";
import { ProcessorNotifier } from "./processors/shared";
import { SlackNotifier } from "../notifiers/slackNotifier";
import { TelegramNotifier } from "../notifiers/telegramNotifier";
import { createIssueBranchName } from "../utils/branch";
import { ForegroundRunner } from "./foregroundRunner";
import { LockFileManager } from "./lockFileManager";

interface ExtendedIssue extends GitHubIssue {
  repo_name?: string;
}

function isSlackNotifier(notifier: ProcessorNotifier): notifier is SlackNotifier {
  return notifier instanceof SlackNotifier;
}

function formatTitle(displayName: string, title: string): string {
  return `[${displayName}] ${title}`;
}

interface ProcessorDefinition {
  name: string;
  displayName: string;
  labels: {
    working: string;
    completed: string;
    failed: string;
  };
  runner: {
    processIssue(
      issueNumber: number,
      agentIndex: number,
      stateManager: StateManager,
      repoName?: string
    ): Promise<{ status: ProcessStatus; sessionId?: string | null }>;
  };
}

export class ImploidOrchestrator {
  private readonly stateManager = new StateManager();
  private readonly githubClient: GitHubClient;
  private readonly repoManager: RepoManager;
  private readonly notifiers: ProcessorNotifier[] = [];
  private readonly processors: ProcessorDefinition[];

  constructor(private readonly config: Config, processorsOverride?: ProcessorName[]) {
    this.githubClient = new GitHubClient(this.config.githubToken);

    if (this.config.telegramBotToken && this.config.telegramChatId) {
      this.notifiers.push(new TelegramNotifier(this.config.telegramBotToken, this.config.telegramChatId));
    }
    if (this.config.slackBotToken && this.config.slackChannelId) {
      this.notifiers.push(new SlackNotifier(this.config.slackBotToken, this.config.slackChannelId));
    }

    this.repoManager = new RepoManager(this.config);

    const definitions: Record<ProcessorName, () => ProcessorDefinition> = {
      claude: () => ({
        name: "claude",
        displayName: "Claude",
        labels: {
          working: "claude-working",
          completed: "claude-completed",
          failed: "claude-failed",
        },
        runner: new ClaudeProcessor(this.config, this.notifiers, this.repoManager),
      }),
      codex: () => ({
        name: "codex",
        displayName: "Codex",
        labels: {
          working: "codex-working",
          completed: "codex-completed",
          failed: "codex-failed",
        },
        runner: new CodexProcessor(this.config, this.notifiers, this.repoManager),
      }),
    };

    const activeNames = processorsOverride && processorsOverride.length
      ? Array.from(new Set(processorsOverride))
      : this.config.enabledProcessors;
    const activeSet = new Set<ProcessorName>(activeNames);

    this.processors = SUPPORTED_PROCESSORS.filter((name) => activeSet.has(name)).map((name) => definitions[name]());

    if (!this.processors.length) {
      throw new Error(
        "No processors enabled. Update your configuration or pass --processors=<names> when running imploid."
      );
    }
  }

  private async ensureState(): Promise<void> {
    await this.stateManager.initialize();
  }

  private async ensureBasePaths(): Promise<void> {
    for (const repoConfig of this.config.githubRepos) {
      const basePath = resolve(repoConfig.base_repo_path.replace(/^~\//, `${process.env.HOME ?? ""}/`));
      await mkdir(basePath, { recursive: true });
      for (const processor of this.processors) {
        await mkdir(resolve(basePath, processor.name), { recursive: true });
      }
    }
  }

  async run(): Promise<void> {
    await this.ensureState();
    await this.ensureBasePaths();

    const allIssues: ExtendedIssue[] = [];
    for (const repoConfig of this.config.githubRepos) {
      try {
        const issues = await this.githubClient.getReadyIssues(repoConfig.name);
        issues.forEach((issue) => {
          (issue as ExtendedIssue).repo_name = repoConfig.name;
        });
        console.info(`Found ${issues.length} ready issues in ${repoConfig.name}`);
        allIssues.push(...(issues as ExtendedIssue[]));
      } catch (error) {
        console.error(`Failed to get issues from ${repoConfig.name}`, error);
      }
    }

    console.info(`Found ${allIssues.length} total ready issues`);

    await this.scheduleIssues(allIssues);
    await this.stateManager.saveStates();
  }

  private async scheduleIssues(issues: ExtendedIssue[]): Promise<void> {
    const activeIssueNumbers = new Set(this.stateManager.getActiveIssueNumbers());
    let remainingCapacity = this.config.maxConcurrent - activeIssueNumbers.size;

    if (remainingCapacity <= 0) {
      return;
    }

    const candidateIssues = issues.filter((issue) => !activeIssueNumbers.has(issue.number));
    if (!candidateIssues.length) {
      return;
    }

    const startPromises: Promise<void>[] = [];

    for (const issue of candidateIssues) {
      if (remainingCapacity <= 0) {
        break;
      }

      const allocations = await this.reserveIssueSlots(issue);
      if (!allocations) {
        console.warn(`Unable to schedule issue #${issue.number}; no available agent slots for all processors`);
        continue;
      }

      activeIssueNumbers.add(issue.number);
      remainingCapacity -= 1;
      startPromises.push(this.processIssueAcrossProcessors(issue, allocations));
    }

    await Promise.all(startPromises);
  }

  private async reserveIssueSlots(
    issue: ExtendedIssue
  ): Promise<Array<{ processor: ProcessorDefinition; agentIndex: number }> | null> {
    const allocations: Array<{ processor: ProcessorDefinition; agentIndex: number }> = [];

    for (const processor of this.processors) {
      const agentIndex = this.stateManager.getAvailableAgentIndex(processor.name, this.config.maxConcurrent);
      if (agentIndex === undefined) {
        return null;
      }
      allocations.push({ processor, agentIndex });
    }

    const repoName = issue.repo_name ?? this.config.githubRepo;
    for (const { processor, agentIndex } of allocations) {
      const branchName = createIssueBranchName(issue.number, processor.name);
      const state = new IssueState({
        issue_number: issue.number,
        status: ProcessStatus.Running,
        branch: branchName,
        start_time: new Date().toISOString(),
        agent_index: agentIndex,
        repo_name: repoName,
        processor_name: processor.name,
        session_id: null,
      });
      this.stateManager.setState(issue.number, processor.name, state);
    }

    await this.stateManager.saveStates();
    return allocations;
  }

  private async processIssueAcrossProcessors(
    issue: ExtendedIssue,
    allocations: Array<{ processor: ProcessorDefinition; agentIndex: number }>
  ): Promise<void> {
    await Promise.all(
      allocations.map(({ processor, agentIndex }) => this.processIssueForProcessor(processor, issue, agentIndex))
    );
  }

  private async processIssueForProcessor(
    processor: ProcessorDefinition,
    issue: ExtendedIssue,
    agentIndex: number
  ): Promise<void> {
    const repoName = issue.repo_name ?? this.config.githubRepo;
    let state = this.stateManager.getState(issue.number, processor.name);
    if (!state) {
      console.error(`State not found for issue #${issue.number} and processor ${processor.name}`);
      return;
    }

    try {
      await this.githubClient.updateIssueLabels(issue.number, {
        add: [processor.labels.working],
        remove: ["agent-ready", processor.labels.completed, processor.labels.failed],
      }, repoName);

      const formattedTitle = formatTitle(processor.displayName, issue.title);
      await Promise.all(
        this.notifiers.map((notifier) =>
          isSlackNotifier(notifier)
            ? notifier.notifyStart(issue.number, formattedTitle, repoName)
            : notifier.notifyStart(issue.number, formattedTitle)
        )
      );

      const result = await processor.runner.processIssue(issue.number, agentIndex, this.stateManager, repoName);
      state = this.stateManager.getState(issue.number, processor.name);

      if (state) {
        if (result.sessionId) {
          state.session_id = result.sessionId;
        }
        state.status = result.status;
        state.end_time = new Date().toISOString();
        this.stateManager.setState(issue.number, processor.name, state);
        await this.stateManager.saveStates();
      }

      if (result.status === ProcessStatus.Completed && state?.start_time && state?.end_time) {
        const start = new Date(state.start_time);
        const end = new Date(state.end_time);
        const durationSeconds = Math.max(0, Math.round((end.getTime() - start.getTime()) / 1000));
        const duration = `${Math.floor(durationSeconds / 60)}m ${durationSeconds % 60}s`;

        await Promise.all(
          this.notifiers.map((notifier) =>
            notifier.notifyComplete(issue.number, duration, processor.displayName, repoName)
          )
        );

        await this.githubClient.updateIssueLabels(issue.number, {
          add: [processor.labels.completed],
          remove: [processor.labels.working],
        }, repoName);

        this.stateManager.removeState(issue.number, processor.name);
        await this.stateManager.saveStates();
      } else if (result.status === ProcessStatus.NeedsInput && state?.last_output) {
        await Promise.all(
          this.notifiers.map((notifier) =>
            isSlackNotifier(notifier)
              ? notifier.notifyNeedsInput(issue.number, state.last_output ?? "", repoName)
              : notifier.notifyNeedsInput(issue.number, state.last_output ?? "")
          )
        );
      } else if (result.status === ProcessStatus.Failed) {
        await this.githubClient.updateIssueLabels(issue.number, {
          add: [processor.labels.failed],
          remove: [processor.labels.working, "agent-ready"],
        }, repoName);
        this.stateManager.removeState(issue.number, processor.name);
        await this.stateManager.saveStates();
      }
    } catch (error) {
      console.error(`Error processing issue #${issue.number} with ${processor.displayName}`, error);
      await this.githubClient.updateIssueLabels(issue.number, {
        add: [processor.labels.failed],
        remove: [processor.labels.working, "agent-ready"],
      }, repoName).catch(() => undefined);
      this.stateManager.removeState(issue.number, processor.name);
      await this.stateManager.saveStates();
    } finally {
      await this.stateManager.saveStates();
    }
  }
}

export interface OrchestratorOptions {
  processors?: ProcessorName[];
  quiet?: boolean;
  foreground?: boolean;
  version?: string;
  description?: string;
}

export async function main(options: OrchestratorOptions = {}): Promise<void> {
  const config = await Config.loadOrCreate();

  if (!options.quiet) {
    const bannerLines = buildStartupBanner(config, {
      version: options.version,
      description: options.description,
      processorsOverride: options.processors,
    });
    for (const line of bannerLines) {
      console.log(line);
    }
  }

  const orchestrator = new ImploidOrchestrator(config, options.processors);

  if (options.foreground) {
    const lockManager = new LockFileManager();
    if (await lockManager.isLocked()) {
      const holder = await lockManager.getCurrentHolder();
      console.error(`\nError: Another Imploid instance is already running in foreground mode`);
      if (holder) {
        console.error(`  PID: ${holder.pid}`);
        console.error(`  Started: ${holder.startTime}`);
      }
      console.error(`\nTo stop the running instance, use: kill ${holder?.pid || '<PID>'}`);
      process.exitCode = 1;
      return;
    }

    const runner = new ForegroundRunner(
      async () => await orchestrator.run(),
      { pollingInterval: 60000 }
    );
    await runner.start();
  } else {
    const lockManager = new LockFileManager();
    if (await lockManager.isLocked()) {
      const holder = await lockManager.getCurrentHolder();
      console.error(`\nError: Imploid is running in foreground mode`);
      if (holder) {
        console.error(`  PID: ${holder.pid}`);
        console.error(`  Started: ${holder.startTime}`);
      }
      console.error(`\nCannot run one-shot mode while foreground mode is active.`);
      console.error(`To stop the foreground instance, use: kill ${holder?.pid || '<PID>'}`);
      process.exitCode = 1;
      return;
    }

    await orchestrator.run();
  }
}
